src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -534,44 +526,88 @@ private static async Task WriteDataAsync(Stream archiveStream, Stream dataStream
             cancellationToken.ThrowIfCancellationRequested();
 
             await dataStream.CopyToAsync(archiveStream, cancellationToken).ConfigureAwait(false); // The data gets copied from the current position
+
             int paddingAfterData = TarHelpers.CalculatePadding(actualLength);
-            await archiveStream.WriteAsync(new byte[paddingAfterData], cancellationToken).ConfigureAwait(false);
+            if (paddingAfterData != 0)
+            {
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(paddingAfterData);
+                Array.Clear(buffer, 0, paddingAfterData);
+
+                await archiveStream.WriteAsync(buffer.AsMemory(0, paddingAfterData), cancellationToken).ConfigureAwait(false);
+
+                ArrayPool<byte>.Shared.Return(buffer);
+            }
         }
 
         // Dumps into the archive stream an extended attribute entry containing metadata of the entry it precedes.
         private static Stream? GenerateExtendedAttributesDataStream(Dictionary<string, string> extendedAttributes)
         {
             MemoryStream? dataStream = null;
+
+            byte[]? buffer = null;
+            Span<byte> span = stackalloc byte[512];
+
             if (extendedAttributes.Count > 0)
             {
                 dataStream = new MemoryStream();
+
                 foreach ((string attribute, string value) in extendedAttributes)
                 {
-                    byte[] entryBytes = GenerateExtendedAttributeKeyValuePairAsByteArray(Encoding.UTF8.GetBytes(attribute), Encoding.UTF8.GetBytes(value));
-                    dataStream.Write(entryBytes);
+                    // Generates an extended attribute key value pair string saved into a byte array, following the ISO/IEC 10646-1:2000 standard UTF-8 encoding format.
+                    // https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html
+
+                    // The format is:
+                    //     "XX attribute=value\n"
+                    // where "XX" is the number of characters in the entry, including those required for the count itself.
+                    int length = 3 + Encoding.UTF8.GetByteCount(attribute) + Encoding.UTF8.GetByteCount(value);
+                    length += CountDigits(length);
```

Comment:
I think this logic needs to be tweaked,  and it should be more like:
```C#
int digitCount = CountDigits(length) ;
length += digitCount;
length += CountDigits(length) - digitCount; // account for possible digit length increase
```

Do we have tests for this stuff?  Do we validate that the archives at produce are readable by other tools?  Our own code appears to ignore this length when reading archives (maybe it shouldn't?)

Reply:
Are we even sure the length is supposed to include itself?  Have you seeb that being done in archives produced by other tools?

Reply:
From [man tar 5](https://manpages.org/tar/5):

```
The extended attributes themselves are stored as a series of text-format lines encoded in the portable UTF-8 encoding. Each line consists of a decimal number, a space, a key string, an equals sign, a value string, and a new line. The decimal number indicates the length of the entire line, including the initial length field and the trailing newline. An example of such a field is:

25 ctime=1084839148.1212\n
```

Reply:
I agree it's strange design.

I did verify that the `tar` tool was able to read our archives containing extended attributes entries.

Since we know the length of the data section, I didn't feel it was too important to verify that the length of each extended attribute entry was correct, considering that we need to look for a mandatory newline char suffix. Maybe if we checked the length number and advanced the position instead of searching for the newline, we could improve perf a bit. I can investigate if this is true and determine if we need to submit a PR to verify the length number.